feat(donate): additional fields on the form

adekbadek opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Adds additional fields feature to the streamlined Donate block:

<img width="810" alt="image" src="https://user-images.githubusercontent.com/7383192/207069863-8e74ddc7-faaa-492c-be16-b5830b9906b8.png">

_(address and phone fields above are added per-block)_

### How to test the changes in this Pull Request:

1. Switch `newspack-plugin` to `feat/donation-additional-fields` branch (https://github.com/Automattic/newspack-plugin/pull/2187)
1. Ensure the site is set up to use Stripe as the Reader Revenue platform and webhooks can reach the site
2. Insert a Donate block on a page. Observe new editor sidebar panel – "Additional data fields"
3. Add some fields
4. Visit the front-end and make a donation, supply values to the additional fields
5. Observe that the fields are added to the Stripe customer metadata – as separate fields, but also the whole group, JSON-encoded, under `_additional_fields` meta key

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+
+type EditableKey = keyof Omit< AdditionalField, 'type' >;
+const FIELD_PROPS = [
+	[ 'label', __( 'Label', 'newspack' ) ],
```
Text domain?

```suggestion
	[ 'label', __( 'Label', 'newspack-blocks' ) ],
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+	[ 'label', __( 'Label', 'newspack' ) ],
+	[
+		'name',
+		__( 'Name', 'newspack' ),
```
```suggestion
		__( 'Name', 'newspack-blocks' ),
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+		__( 'Name', 'newspack' ),
+		__(
+			'Name of the field which will be sent to the payment procesor and other third parties. Field names must be unique.',
+			'newspack'
```
```suggestion
			'newspack-blocks'
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+			<Button
+				onClick={ toggleVisible }
+				icon={ moreVertical }
+				label={ __( 'Options', 'newspack' ) }
```
```suggestion
				label={ __( 'Options', 'newspack-blocks' ) }
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+					className={ `${ BASE_CSS_CLASSNAME }__options-popover` }
+				>
+					<MenuItem onClick={ onEdit } isLink>
+						{ __( 'Edit', 'newspack' ) }
```
```suggestion
						{ __( 'Edit', 'newspack-blocks' ) }
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+						} }
+						isLink
+					>
+						{ __( 'Remove', 'newspack' ) }
```
```suggestion
						{ __( 'Remove', 'newspack-blocks' ) }
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+					attributes.additionalFields.filter(
+						( { name }, i ) => i !== field.fieldIndex && name === field.name
+					).length === 0;
+				return ! isValid ? __( 'Name already exists.', 'newspack' ) : '';
```
```suggestion
				return ! isValid ? __( 'Name already exists.', 'newspack-blocks' ) : '';
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+				);
+			} ) }
+			<ToggleControl
+				label={ __( 'Required', 'newspack' ) }
```
```suggestion
				label={ __( 'Required', 'newspack-blocks' ) }
```

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,240 @@
+						setEditedField( newField );
+					} }
+				>
+					{ __( 'Add data field', 'newspack' ) }
```
```suggestion
					{ __( 'Add data field', 'newspack-blocks' ) }
```

dkoo commented on src/blocks/donate/edit/index.tsx:
```diff
@@ -30,28 +30,22 @@ import { isEmpty, pick } from 'lodash';
-};
-
 const TIER_LABELS = [
 	__( 'Low-tier', 'newspack' ),
```
These two, also, even though they're already in `master`

dkoo reviewed (changes requested):
This works really well. Saw a few instances of the wrong text domain, below, and I also have few feature requests, the first more important and the others not so important (non-blocking).

1. Can we implement some way to reorder the fields? Right now the only way to reorder them is to delete and rebuild. Some up/down arrow icons on the field buttons would be enough, IMO.
2. (Non-blocking) The fields all get rendered at full-width, which can look awkward on a large screen. Maybe we could add an option to the custom fields to make them full-width or half-width? This will be a fairly common set of fields, and State and ZIP could definitely be half-width at larger viewports:

<img width="1201" alt="Screen Shot 2022-12-16 at 2 14 05 PM" src="https://user-images.githubusercontent.com/2230142/208209358-109ca2c9-5d30-44d7-8239-0453b5a2ca17.png">

3. (Non-blocking) A nitpick, but can we make the field names in the sidebar clickable to open the edit modal? They look like buttons, so I want to click them! 😄 

<img width="264" alt="Screen Shot 2022-12-16 at 4 50 50 PM" src="https://user-images.githubusercontent.com/2230142/208209427-7da87f71-d4cc-40fb-a408-f0ebab72bd90.png">

adekbadek commented:
Fixed the text domain wholesale in 76c4e5a

> Can we implement some way to reorder the fields? 

Great idea! Done in 5a3fcf9

>  A nitpick, but can we make the field names in the sidebar clickable to open the edit modal? T

Done in d33238f. FYI @thomasguillot here's the updated UI:

<img width="302" alt="image" src="https://user-images.githubusercontent.com/7383192/208405171-4d5329dd-fb47-4a79-b6fe-5f4dd6451e71.png">

> The fields all get rendered at full-width, which can look awkward on a large screen. 

Added a width setting:

<img width="413" alt="image" src="https://user-images.githubusercontent.com/7383192/208412642-c6aa63a7-40de-4d3a-83da-7557bd4797ea.png">

Let me know if the styling should be tweaked @thomasguillot

adekbadek commented:
Update: reverted field label clickability as per @thomasguillot 's request, in b61bc3c.

dkoo commented on src/blocks/donate/frontend/class-newspack-blocks-donate-renderer-base.php:
```diff
@@ -246,6 +246,20 @@ protected static function render_streamlined_payment_ui( $attributes ) {
+								type="<?php echo esc_attr( $field['type'] ); ?>"
+								name="<?php echo esc_attr( $field['name'] ); ?>"
+								placeholder="<?php echo esc_attr( $field['label'] ); ?>"
+								style="width: calc(<?php echo esc_attr( $field['width'] ); ?>% - 0.5rem);"
```
This probably only happened to me because I had fields created before you added the `width` property, but I saw a fatal at first for fields that didn't have a set width. The fatal did not happen after I deleted and recreated the fields.

```suggestion
								<?php if ( isset( $field['width'] ) ) : ?>
								style="width: calc(<?php echo esc_attr( $field['width'] ); ?>% - 0.5rem);"
								<?php endif; ?>
```

adekbadek replied:
Yep, that's expected. The feature is not merged yet so there's no need to handle "legacy" fields.

dkoo reviewed (changes requested):
Thanks for the changes! I think the widths and reordering add a lot of value to this feature.

I ran into some new minor issues with the widths:

1. The widths don't quite add up to 100%. Here's what it looks like if I set three fields with 33% width:

<img width="1221" alt="Screen Shot 2022-12-19 at 4 11 54 PM" src="https://user-images.githubusercontent.com/2230142/208546868-e9d86001-f5c8-4112-b31d-694ef8f8a643.png">

2. This is another feature request, sorry. Is it possible to add non-interactive placeholder fields to the editor preview of the block? I found myself having to view the post in a preview tab and switch back and forth between the two in order to see what my changes would look like, which seems like something publisher's won't want to do.

3. Non-blocking: I found myself wanting a `25%` option for the widths, too, so I could set a row to 50%, 25%, 25%. Too much?

4. Non-blocking: This may only be required if you introduce a `25%` option, but I think the widths could potentially be applied only to larger viewports (so for mobile they'd all be 100%). We'd need some help text to explain this if we go this route.

gamebits commented:
Can these fields be set to invisible? For example, if there is static data the publisher wants sent to Stripe with every transaction, to distinguish Newspack donations from other Stripe activity. (Or is there another way to do that?)

adekbadek commented:
That's an interesting feature request – can you think of any other uses besides marking transaction as Newspack-originating? We're already adding a `current_page_url` and some other metadata by default, this field could be another one of these. 

> Or is there another way to do that?

I don't think so.

adekbadek commented:
> The widths don't quite add up to 100%. Here's what it looks like if I set three fields with 33% width:

Fixed in 4fc0b9

> Is it possible to add non-interactive placeholder fields to the editor preview of the block? 

Yes! Great suggestion – added in c19249a8cb9a4bc7ce544475aefe8b312822c275

> Non-blocking: I found myself wanting a 25% option for the widths, too, so I could set a row to 50%, 25%, 25%. Too much?

I did ask @thomasguillot in a DM previously and he said it's too much

dkoo commented on src/blocks/donate/edit/components/AdditionalFields.tsx:
```diff
@@ -0,0 +1,282 @@
+			<div className={ `${ BASE_CSS_CLASSNAME }__field-edited__width` }>
+				<div>{ __( 'Field width:', 'newspack-blocks' ) }</div>
+				<ButtonGroup>
+					{ [ 100, 66, 50, 33.33 ].map( width => (
```
```suggestion
					{ [ 100, 66.66, 50, 33.33 ].map( width => (
```

leogermani replied:
> (e.g. if you have a 66% and 33% field side-by-side, they should add up to 100%).

Shouldn't it be 66.67 or 33.34 then?

dkoo replied:
The `Math.round` [here](https://github.com/Automattic/newspack-blocks/pull/1330/files#diff-049b01968454fddef70955d3fd9e376175da7a45a99ca02cfbe7ae017c73d2ccR155) will ensure that it's rounded up to 100% when totaled.

dkoo reviewed (approved):
Thanks for the preview feature—it makes editing the block much more user-friendly! Pre-approving because there's just one small fix needed to ensure that 66% width fields are rendered at the full width (e.g. if you have a 66% and 33% field side-by-side, they should add up to 100%).
